### ISSUE TITLE
no-try-catch rule?

### ISSUE THREAD

**Issue by malandrew** [eslint/eslint#2422]
try-catch statements can be problematic in async code on the server. 

It's generally best for developers to identify any source of thrown errors in libraries they consume that should be caught and handled and make a small wrapper that makes it safe to use that library everywhere without worrying about throwing. This could possibly include converting a thrown error into a NodeJS style callback. 

I propose a rule that is off-by-default (maybe on by default for nodejs env?) that warns when you use try-catch, leading you to explicitly disable the rule with a directive when you deliberately want to use try-catch.

A PR for this already exists here https://github.com/eslint/eslint/pull/2421 but we can discuss the rule in this issue.

Thoughts?

**Comment by Raynos** [comment]
:+1:

Although it should never be true by default.

This is really a callbacks vs promises thing. When doing callback driven nodejs any `try {} catch {}` is the devil for reasoning about errors (not to mention performance).

When doing promises; it's actually recommended to use `try {} catch {}`

This is a useful feature to have to enforce correctness in nodejs programs.

**Comment by nzakas** [comment]
I'm not sure if this fits with the bundled rules in ESLint right now. We don't have a lot of "don't use this language feature" rules. I need some convincing that this is common enough that it merits inclusion vs. you publishing as a plugin.

@ilyavolodin @btmills @lo1tuma @gyandeeps thoughts?

**Comment by ilyavolodin** [comment]
Arguably `no-with` also disables language feature, however, while I can see not wanting to use `try-catch` for browser code that has to support older browsers (I'm not really sure I follow Node callback example), I feel like this might be too situational of a rule to be in the core, and would be better off being a plugin.

**Comment by malandrew** [comment]
Basically, you never want to throw errors in async code. You always want to return the error to the caller for the caller to decide what to do. The caller may want to handle the error and continue or the caller may decide to throw, but the decision to throw should always be delegated to the caller. Basically, if you're writing nodejs libraries that are async you shouldn't throw ever. Here is izs (maintainer of nodejs and original author of npm) on throwing in async code:

https://groups.google.com/forum/#!msg/nodejs/W9UVJCKcJ7Q/rzseRbourCUJ

> "Throwing in async code is sloppy."

If eslint is designed to prevent sloppy code, and most nodejs code is async, and throwing is sloppy, then this should be a common enough rule that it should be in the core. The entire nodejs ecosystem would be better off if you could trust all async libraries to never throw. You have no idea how much trouble the apache/thrift nodejs code is to deal with because the "never throw" rule wasn't followed. It's never an appropriate flow control pattern in async code.

In synchronous code, however, try-catch is fine.

**Comment by ilyavolodin** [comment]
@malandrew Thanks for the detail explanation, now it makes sense. However, if that's the case, I would argue that we should instead create a rule called `no-throw` or `no-throw-inside-async`. You argument is to never throw in async code. But if I have to use 3-rd party library that throws inside my async code (might not have a choice), then I have to use try/catch in order to follow your own advice.

**Comment by gcochard** [comment]
You might also run into issues where there are core node APIs that are both sync and async like child_process.exec (joyent/node#8573).

**Comment by malandrew** [comment]
@ilyavolodin While I love that idea, I know of no way to determine that code is async or not via static analysis with the AST. If you have any suggestions, I'd be happy to explore that idea.

For the case where you "have to use 3-rd party library that throws inside my async code (might not have a choice)", we typically use the `/*eslint-disable no-try-catch*/ ... code ... /*eslint-enable no-try-catch*/` directives. This is what is suggested in the docs I added for this rule in my pull request:
https://github.com/eslint/eslint/pull/2421/files#diff-9d1052a3cbb1cbd7a0c0821f9d46acb5R23

The use of try/catch should be extremely rare and the exception, not the rule. It's just like with max-statements or max-line length. 99% of the time you probably want max-statements or max-line-length on, but every once in a while you disable that rule infrequently and only where it makes sense to do so.

The only feature that unfortunately became a standard in the language that I can think of that throws where you would want to catch the error and not crash the process is JSON.parse and JSON.stringify. For that, I always recommend that people use safe-json-parse or write their own little helper function for doing the same with `eslint-disable` directives around that helper function.

Whenever I encounter a library that throws in async code, I take that as a sign that I should go ahead and look around for a replacement or write one that does the right thing.

@gcochard how unfortunate that exec throws. That should be fixed. You should never have to deal with both throwing and a callback. What were you doing that you discovered that issue?

**Comment by lo1tuma** [comment]
I don’t understand why you want a `no-try-catch` rule if your only argument is to avoid throwing in async code. Are you sure that you don’t want a `no-throw` rule instead, as @ilyavolodin suggested? If an async function never throws it should’t cause any problems calling it within a `try` block.

**Comment by nzakas** [comment]
Reminder: we have `no-with` because `with` is both deprecated and not allowed in strict mode, so not a direct analogy here.

I think @lo1tuma and @ilyavolodin hit the nail on the head. It seems like what you're really after is `no-throw`. Even so, I'm so not sure we want to start adding one-off rules disabling language features. If there's a need, we should probably just create a rule that lets you flag any arbitrary AST node and be done with it.

I'd also point out that you could just as well take this code and publish it as a plugin. Part of why we have plugins is so we don't need to be gatekeepers for all rules in the world.

**Comment by malandrew** [comment]
You might still want to throw if your intent is to crash the process, which is appropriate for programmer logic errors as opposed to operational errors. What I'm trying to prevent is the usage of try/catch for handling programmer errors. If you know enough to try/catch a programmer error, then you know enough to prevent the error from ever being thrown in the first place.

instead of this:

```
function getFoo(value, callback) {
    callback = once(callback);
    try {
        unsafeAsyncFetchFoo(value, callback);
    } catch(e) {
        process.nextTick(function() {
            callback(e);
        });
    }
}
```

prefer this:
```
function getFoo(value, callback) {
    // If you know something could throw, enforce a contract by checking a value before passing
    // it to a function that could throw.
    if (typeof value !== 'string') {
        return process.nextTick(function() {
            callback(new Error('value must be a string'));
        });
    }
    unsafeAsyncFetchFoo(value, callback);
}
```

https://www.joyent.com/developers/node/design/errors

**Comment by Raynos** [comment]
@lo1tuma @ilyavolodin 

> create a rule called no-throw or no-throw-inside-async. You argument is to never throw in async code.

Throwing is perfectly acceptable. You want to throw for programming exceptions. For example

```js
function readFile(path, encoding, callback) {
  if (typeof path !== 'string') {
    throw new Error('expected path to be a string');
  }
}
```

or alternatively

```js
WritableStream.prototype._write = function () {
  throw new Error('_write is not implemented in base class');
};
```

All thrown exceptions are fatal. What we want to discourage is people using exceptions for non-fatal purposes; and we ban `try {} catch {}` for that reason.

If your using try catch; you are also using `throw` for non fatal use cases. It's really hard to ban `throw` for the non-fatal cases because you cannot detect whether a throw is fatal or not.

**Comment by ilyavolodin** [comment]
Hmm.. OK, I think I can buy that argument. I still think it's a bit situational rule, as it doesn't really address the actual problem, but I can see how it might address 80% of the cases that might cause issues. So I think I'm OK with this rule, as long as it's off by default.

**Comment by gcochard** [comment]
@malandrew lots of subprocesses in an internal web app for generating docs. Like 4 per POST and the front end had a batch mode and many users on at once.

Back on topic, I agree that this should be considered for a first class rule. The exceptions to this rule are for sync APIs that abuse `throw` for non-critical errors, or for that matter easy to handle conditions.

**Comment by nzakas** [comment]
To reassert my previous point, I'm not a fan of rules that disallow single language features like this (absent evidence, such as `with`). If we add `no-try-catch`, then we would open the door for `no-throw`, `no-switch`, `no-do-while`, and we will get caught in an avalanche of one-off rules.

If we want to add something to support this case, then I believe it must be a generic `no-restricted-syntax` that lets you specify which AST nodes should cause a warning.

**Comment by glenjamin** [comment]
+1 to not having this in core, it seems perfectly sensible to live in a plugin.

This is a (somewhat) common pattern to appear in async code for example:

```js
function getSomeData(id, callback) {
  api.getJSONString(id, function(err, json) {
    if (err) return callback(err);
    try {
      return callback(null, JSON.parse(json));
    } catch (ex) {
      return callback(ex);
    }
  });
}
```

I know this sort of thing can be packaged up into helper functions, but the pattern will likely appear wherever there is a seam between exception-throwing synchronous code, and a call to it from async code.

**Comment by malandrew** [comment]
@nzakas 

Out of curiosity, how does this proposed rule differ from rules like `no-ternary` and `no-plusplus` with respect to this statement " I'm not a fan of rules that disallow single language features like this". That rule very specifically disallows a language feature that is more benign that try/catch for async code (which is a large portion of all the JS code out in the wild).

https://github.com/eslint/eslint/blob/master/docs/rules/no-ternary.md

I'm more just curious why. I'm pretty sure the answer will still be no (in which case, feel free to go ahead and close this)

**Comment by nzakas** [comment]
`no-plusplus` was for compatibility with JSLint. 

`no-ternary` was introduced very early on (0.0.9), and we hadn't yet figured out what things made sense and what didn't. Also, there was no way to write custom rules at that point, so there were no options outside of supporting as many rules as we could.

**Comment by gyandeeps** [comment]
@eslint/eslint-team feedback/suggestions/thoughts?

**Comment by BYK** [comment]
:-1:

**Comment by nzakas** [comment]
We get enough requests like this that I think we need a generic approach.

I'd like to propose a `no-restricted-syntax` rule such that you can provide AST node types that are not allowed in your code. For example:

```
no-restricted-syntax: [2, "TryStatement"]
```

Then we just look for the nodes in the question and warn if we find them. That let's anyone restrict anything at any point in time and we don't have to keep thinking about adding restrictions for each type of syntax.

Thoughts?

**Comment by gyandeeps** [comment]
On @nzakas proposal: :+1:

**Comment by lo1tuma** [comment]
FWIW there is already a plugin which allows you to disable language features: [`eslint-plugin-better`](https://github.com/idmitriev/eslint-plugin-better).

**Comment by BYK** [comment]
:+1: to @nzakas' proposal. The mentioned plugin looks good too though, although it doesn't follow the suggestion.

**Comment by nzakas** [comment]
OK, let's do this.

**Comment by scottohara** [comment]
Was just trying out this new rule on a codebase that uses ES6 classes (`parser: "babel-eslint"`).

With `FunctionExpression` in the list of restrictions (as per the docs), I get  `error  Using "FunctionExpression" is not allowed` on the class constructor, e.g.:

```
/* eslint no-restricted-syntax: [2, "FunctionExpression"] */

class Foo {
  constructor() {
    this.bar = "baz";
  }
}
```

Presumably this is because `babel` compiles ES6 class syntax to the following ES5 code:

```
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Foo = function Foo() {        // <- function expression here
  _classCallCheck(this, Foo);

  this.bar = "baz";
};
```

Is it expected that this rule (as configured) should treat ES6 classes as an invalid use of `FunctionExpression`?

If not, would this be an issue for `babel-eslint` or `eslint` to address?

**Comment by lo1tuma** [comment]
Can you try it with the default parser? If that works correctly then this is a problem with babel eslint.

**Comment by scottohara** [comment]
Same error with `parser: "espree"`

**Comment by mysticatea** [comment]
I think that is an expected behavior.
eslint is using JavaScript AST.
This rule is to forbid specified AST node types.
And MethodDefinition nodes have a FunctionExpression node in its value.

**Comment by scottohara** [comment]
OK, thanks @mysticatea .  Just wasn't sure if it was expected or not.
